File: 2019/day-2/sol.js
// setup {{{
let aocInput;
{
  const fs = require('fs');
  aocInput = fs.readFileSync('input', 'utf-8');
}
const test = (fn, samples) => {
  for (const i in samples) {
    const [arg, expectedResult] = samples[i];
    const actualResult = fn(arg, false);

    const strExpected = JSON.stringify(expectedResult);
    const strActual = JSON.stringify(actualResult);

    if (strExpected === strActual) {
      console.log(`test ${i}: ok (${strActual})`);
    } else {
      console.log();
      console.group();
      console.group();
      console.log(expectedResult);
      console.groupEnd();
      console.log('instead got v');
      console.group();
      console.log(actualResult);
      console.groupEnd();
      console.groupEnd();
    }
  }
};
// }}}

const processInput = input => input.trim().split(',').map(Number);

const sol1 = (input, real=true, d1=12, d2=2) => {
  const ds = processInput(input);

  if (real) {
    ds[1] = d1;
    ds[2] = d2;
  }

  let sec = 10000;
  let i = 0;
  while (--sec > 0) {
    const v = ds[i];

    if (v === 1) {
      const [a, b, res] = [ds[i + 1], ds[i + 2], ds[i + 3]];
      ds[res] = ds[a] + ds[b];
      i += 4;
    } else if (v === 2) {
      const [a, b, res] = [ds[i + 1], ds[i + 2], ds[i + 3]];
      ds[res] = ds[a] * ds[b];
      i += 4;
    } else if (v === 99) {
      break;
    }
  }

  return ds[0];
};

const sol2 = (input) => {
  for(let i = 0; i < 99; ++i) {
    for(let j = 0; j < 99; ++j) {
      if(sol1(input, true, i, j) === 19690720) {
        return i * 100 + j;
      }
    }
  }
};

const samples1 = [
  ['1,9,10,3,2,3,11,0,99,30,40,50', 3500],
  ['1,0,0,0,99', 2],
  ['2,3,0,3,99', 2],
  ['2,4,4,5,99,0', 2],
  ['1,1,1,4,99,5,6,0,99', 30],
];
const samples2 = [
  [],
];

//test(sol1, samples1);
console.log(sol1(aocInput));

// test (sol2, samples2);
console.log(sol2(aocInput));

File: 2019/day-2/sol.rb
# frozen_string_literal: true

# setup {{{
aoc_input = File.read('input')

def test(fname, samples)
  f = method(fname)
  puts "TESTING #{fname}"
  samples.each_with_index do |sample, i|
    print " test #{i}: "
    res = sample.last
    r = f.call(*sample[0..-2])
    if r != res
      puts 'FAIL '
      puts '   expected'
      puts "     #{res}"
      puts '   but got'
      puts "     #{r}"
    else
      puts 'ok'
    end
  end
end
# }}}

def process_input(input)
  input.strip.split(',').map(&:to_i)
end

def run_intcode(ns)
  i = 0

  while i < ns.size
    op, a, b, res = ns[i..i + 3]

    break if op == 99

    ns[res] = op == 1 ? ns[a] + ns[b] : ns[a] * ns[b]

    i += 4
  end

  ns
end

def sol1(input)
  xs = process_input(input)
  xs[1] = 12
  run_intcode(xs)
  xs[0]
end

def sol2(input)
  xs = process_input(input)
  (0..99).each do |a|
    (0..99).each do |b|
      ns = xs.clone
      ns[1..2] = a, b
      return a * 100 + b if run_intcode(ns)[0] == 19_690_720
    end
  end
end

def test(*_) 'uncomment to disable tests'; end # rubocop:disable SingleLineMethods

test :sol1, [
  []
]
p sol1 aoc_input
test :sol2, [
  []
]
p sol2 aoc_input
